File: app/lib/core/api_client.dart
import 'dart:async';
import 'package:dio/dio.dart';
import 'tokens/token_store.dart';
import 'package:flutter/foundation.dart'; // pra debugPrint

class ApiClient {
  final Dio dio;

  ApiClient._(this.dio);

  static ApiClient create({
    required TokenStore tokenStore,
    required Future<void> Function() onUnauthorized,
    String baseUrl = const String.fromEnvironment(
      'API_BASE_URL',
      defaultValue: 'http://10.0.2.2:5078',
    ),
  }) {
    final dio = Dio(
      BaseOptions(
        baseUrl: baseUrl,
        contentType: 'application/json',
        responseType: ResponseType.json,
        connectTimeout: const Duration(milliseconds: 15000),
        receiveTimeout: const Duration(milliseconds: 20000),
      ),
    );

    bool isRefreshing = false;
    final waiters = <Completer<void>>[];

    dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) {
          final t = tokenStore.token;

          debugPrint('➡️ REQUEST: ${options.method} ${options.uri}');
          debugPrint('   Headers antes: ${options.headers}');
          if (options.data != null) {
            debugPrint('   Body: ${options.data}');
          }

          if (t != null && t.isNotEmpty) {
            options.headers['Authorization'] = 'Bearer $t';
            debugPrint('   Authorization: Bearer ${t.substring(0, 10)}...');
          } else {
            debugPrint('   ⚠️ Sem token no TokenStore');
          }

          handler.next(options);
        },
        onResponse: (response, handler) {
          debugPrint('✅ RESPONSE: [${response.statusCode}] ${response.requestOptions.uri}');
          if (response.data != null) {
            debugPrint('   Data: ${response.data}');
          }
          handler.next(response);
        },
        onError: (e, handler) async {
          final status = e.response?.statusCode;
          final path = e.requestOptions.path;

          debugPrint('❌ ERROR: [${status}] ${e.requestOptions.method} ${e.requestOptions.uri}');
          debugPrint('   Path: $path');
          if (e.response?.data != null) {
            debugPrint('   Response data: ${e.response!.data}');
          }

          final isAuthPath =
              path.contains('/auth/login') ||
              path.contains('/auth/refresh') ||
              path.contains('/api/login');

          // --- versão SEM retry automático, só chama onUnauthorized e repassa erro ---
          if (status == 401 && !isAuthPath) {
            debugPrint('   ⚠️ 401 recebido, chamando onUnauthorized()');
            await onUnauthorized();
            // não tenta refazer a requisição, só devolve o erro
            return handler.next(e);
            // ou: return handler.reject(e);
          }

          handler.next(e);
        },
      ),
    );

    return ApiClient._(dio);
  }
}

File: portalweb/src/services/apiClient.js
import axios from 'axios'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:5078/'

let _authToken = null

export function setAuthToken(token) {
  _authToken = token
}

export function clearAuthToken() {
  _authToken = null
}

const api = axios.create({
  baseURL: API_BASE_URL,
  timeout: 15000,
})

api.interceptors.request.use(
  (config) => {
    if (config.useAuth && _authToken) {
      config.headers = config.headers || {}
      config.headers.Authorization = `Bearer ${_authToken}`
    }

    return config
  },
  (error) => Promise.reject(error),
)

function get(url, options = {}) {
  const { params, auth = false, ...rest } = options
  return api.get(url, { params, useAuth: auth, ...rest })
}

function post(url, data, options = {}) {
  const { params, auth = false, ...rest } = options
  return api.post(url, data, { params, useAuth: auth, ...rest })
}

function put(url, data, options = {}) {
  const { params, auth = false, ...rest } = options
  return api.put(url, data, { params, useAuth: auth, ...rest })
}

function patch(url, data, options = {}) {
  const { params, auth = false, ...rest } = options
  return api.patch(url, data, { params, useAuth: auth, ...rest })
}

function del(url, options = {}) {
  const { params, auth = false, ...rest } = options
  return api.delete(url, { params, useAuth: auth, ...rest })
}

export const apiClient = {
  get,
  post,
  put,
  patch,
  delete: del,
}

export default apiClient
